impeller/entity/contents/filters/gaussian_blur_filter_contents.cc

```diff
@@ -300,21 +279,33 @@ std::optional<Entity> GaussianBlurFilterContents::RenderFilter(
   // gutter from the expanded_coverage_hint, we can skip the downsample pass.
   // pass.
   Vector2 downsample_scalar(desired_scalar, desired_scalar);
-  Vector2 padded_size =
-      Vector2(input_snapshot->texture->GetSize()) + 2.0 * padding;
-  Vector2 downsampled_size = padded_size * downsample_scalar;
+  Rect source_rect = Rect::MakeSize(input_snapshot->texture->GetSize());
+  Rect source_rect_padded = source_rect;
+  Matrix padding_snapshot_adjustment;
+  if (!coverage_hint.has_value() ||
+      (input_snapshot_coverage.has_value() &&
+       !input_snapshot_coverage->Contains(coverage_hint.value()))) {
```

Comment:
A couple of things come to mind while thinking about this:
1. Did you mean to compare against the expanded coverage hint instead? At the end of the downsample pass, we need to have enough content in the downsample pass to cover sampling from the entire expanded coverage area area for the full area of the expanded coverage hint.
2. Should this check be flipped? Like if the expanded_coverage_hint is equal to or a subset of the input_snapshot_coverage, then I suppose that would indicate that the snapshot rendered content into a big enough square that we don't need to worry about creating more with the proper tiling mode.

Reply:
I don't think so, the `input_snapshot_coverage` is incorporating the expanded coverage hint.

This is basically saying: I requested an `expanded_coverage_hint` and the snapshot you gave me for that doesn't contain all of the area of the `coverage_hint`.  Therefore, some part of the halo will be visible and we'll add padding.

Reply:
Check out the test, that should make this clear.  The `coverage_hint` is saying we are going to render a small portion of the image, so there is no need to add the halo gutter.

Reply:
Ah, I actually missed the not operator in front of the `Contains` check here.

I still think this isn't quite working for all cases. I made an interactive playground (will push a PR for it) to demonstrate an issue with Decal tiling:

https://github.com/flutter/engine/assets/919017/91257fc1-a04e-4d60-a26d-bc80617f36fa

Reply:
(Made a PR for the interactive toy here: https://github.com/flutter/engine/pull/49283)

Reply:
Ahh right, the coverage_hint has to be inside of the snapshot image coverage while accounting for the blur radius.  I'll fix that although I wonder if that will make this optimization hardly applicable.

Reply:
I fixed the problem, but there is a jump in the rendering because of the downsampling issue.  It also catches way less cases now because it isn't cutting off the padding when hitting the edge of the input.  I'm going to keep all the refactors and land it as a refactor since I have PRs that depend on this.  I can circle back.

I also ran into a fun Rect floating point bug =( https://github.com/flutter/flutter/issues/140464